format-common/src/main/java/io/cdap/plugin/format/connector/AbstractFileConnector.java

```diff
@@ -163,24 +214,75 @@ protected String getFullPath(String path) {
+   * @param fileName name of the file. Used to guess correct format based on extension.
+   */
+  protected void addBrowseSampleDefaultValues(BrowseEntity.Builder entity, String fileName) {
+    entity.addProperty(SAMPLE_FORMAT_KEY, BrowseEntityPropertyValue.builder(
+      FileTypeDetector.detectFileFormat(FileTypeDetector.detectFileType(fileName)).name().toLowerCase(),
+      BrowseEntityPropertyValue.PropertyType.SAMPLE_DEFAULT).build());
+    entity.addProperty(SAMPLE_FILE_ENCODING_KEY, BrowseEntityPropertyValue.builder(
+      SAMPLE_FILE_ENCODING_DEFAULT, BrowseEntityPropertyValue.PropertyType.SAMPLE_DEFAULT).build());
   }
 
-  private ValidatingInputFormat getValidatingInputFormat(ConnectorContext context, String path) throws IOException {
+  /**
+   *
+   * @return Set of available sample properties
+   */
+  protected Set<BrowseEntityTypeInfo> getSampleProperties() {
+    return sampleProperties;
+  }
+
+  /**
+   * Receives a request and returns a map with only the valid sample properties.
+   *
+   * @param request Request containing the sample properties.
+   * @return a Map with chosen sample properties
+   */
+  protected Map<String, String> getSourceSampleProperties(ConnectorSpecRequest request) {
+    Map<String, String> requestProperties = request.getProperties();
+    Map<String, String> sourceProperties = new HashMap<>();
+    if (requestProperties != null) {
+      for (String sampleKey : SAMPLE_FIELD_NAMES) {
+        if (requestProperties.containsKey(sampleKey)) {
+          sourceProperties.put(sampleKey, requestProperties.get(sampleKey).toLowerCase());
+        }
+      }
+    }
+    return sourceProperties;
+  }
+
+  private ValidatingInputFormat getValidatingInputFormat(ConnectorContext context, String path,
+                                                         Map<String, String> sampleProperties) throws IOException {
+    PluginProperties.Builder builder = PluginProperties.builder();
+    builder.addAll(sampleProperties);
+
     String fileType = FileTypeDetector.detectFileType(path);
     if (!FileTypeDetector.isSampleable(fileType)) {
       throw new IllegalArgumentException(String.format("The given path %s cannot be sampled.", path));
     }
 
-    FileFormat format = FileTypeDetector.detectFileFormat(fileType);
-    PluginProperties.Builder builder = PluginProperties.builder();
     builder.add("path", path);
+    FileFormat format;
+    if (sampleProperties.containsKey("format")) {
+      format = FileFormat.valueOf(sampleProperties.get("format").toUpperCase());
+    } else {
+      format = FileTypeDetector.detectFileFormat(fileType);
+      builder.add("format", format.name());
+    }
+
     if (format.equals(FileFormat.TEXT)) {
       builder.add("schema", DEFAULT_TEXT_SCHEMA);
     }
     builder.addAll(config.getProperties().getProperties());
     builder.addAll(getFileSystemProperties(path));
-
+    builder.add("fileSystemProperties", GSON.toJson(getFileSystemProperties(path)));
```

Comment:
Why do we need to add this, isn't file system properties already set in Line 284?

Reply:
added comment